src/main/java/org/elasticsearch/plugins/repository/PluginResolver.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins.repository;
+
+import com.google.common.collect.Sets;
+import org.elasticsearch.common.collect.Tuple;
+
+import java.util.Collection;
+import java.util.Set;
+
+public class PluginResolver {
+
+    private final PluginRepository local;
+    private final PluginRepository remote;
+
+    public PluginResolver(PluginRepository local, PluginRepository remote) {
+        this.local = local;
+        this.remote = remote;
+    }
+
+    public Resolution resolve(PluginDescriptor plugin) {
+        // Resolves the plugin with its dependencies
+        Resolution report = new Resolution();
+        doResolve(plugin, report);
+
+        return report;
+    }
+
+    private void doResolve(PluginDescriptor plugin, Resolution report) {
+        if ((plugin != null) && (report != null)) {
```

Comment:
We started just some lines before with a `new Resolution()` so `report` can not be `null`.